Améliore l'affichage des références bibliographiques

ggrossetie opened the pull request:
Repose sur #400 

Permet de filtrer/rechercher dans la bibliographie.
Par défaut on affiche au maximum 10 résultats : 

![filter-biblio](https://user-images.githubusercontent.com/333276/120676614-832f8480-c496-11eb-99f1-aa660cac0c73.gif)

Cela améliore drastiquement les performances : 

### Avant

![slow](https://user-images.githubusercontent.com/333276/120677101-05b84400-c497-11eb-8836-22744ff88ea1.gif)

### Après

![fast](https://user-images.githubusercontent.com/333276/120677111-08b33480-c497-11eb-983c-cdf4473a64c6.gif)

Ce n'est pas évident à voir sur le gif car le nombre d'images par seconde est faible mais on passe d'une interface avec des "lags" de 1 à 2 secondes, à un affichage quasiment instantanée.

ggrossetie committed:
Filtre les références bibliographiques et affiche les 10 premiers résultats

ggrossetie commented:
@lakonis @antoinentl Avant d'aller trop loin je serais curieux d'avoir votre avis sur ce principe de fonctionnement.

Les données ne s'enregistrent pas correctement mais c'est surtout pour avoir votre avis sur le composant qui permet de rechercher/filtrer les références bibliographiques (et qui affiche au maximum 10 résultats)

ggrossetie commented:
@antoinentl J'attire ton attention sur ce point qui permet d'améliorer significativement les performances de l'éditeur dès que la bibliographie contient plusieurs centaines (milliers) d'éléments. Est-ce que tu pourrais s'il te plait valider que ce nouveau mode de fonctionnement est adapté ? Merci !

À noter que l'ajout/suppression/modification ne fonctionne pas, pour le moment il faut uniquement valider le nouvel affichage des références bibliographiques (i.e., affichage des 10 premiers éléments avec une barre de recherche/filtre).

antoinentl commented:
@Mogztter Pour résumer :

- il y a rarement plus de 30 références, et très rarement plus de 100 références, mais si cela fait gagner en performance ça vaut le coût d'explorer cette option ;
- afficher seulement une partie de la bibliographie est une option, @Mmellet et @marviro qu'est-ce que vous en pensez ?
- si l'on retient ce principe il faudrait :
    - ne pas limiter à 10 références mais à 20 ou 25 ;
    - indiquer plus clairement que toutes les références ne sont pas affichées ;
    - proposer une option pour _dérouler_ les autres références plutôt que rechercher : je ne sais pas forcément quelles sont mes références, donc je ne sais pas quoi chercher.

ggrossetie commented:
> il y a rarement plus de 30 références, et très rarement plus de 100 références, mais si cela fait gagner en performance ça vaut le coût d'explorer cette option ;

Cela permet surtout de garantir que les performances seront "stables" quelque soit le nombre de références.

> si l'on retient ce principe il faudrait :
ne pas limiter à 10 références mais à 20 ou 25 ;

On peut augmenter un peu la limite mais je pense qu'au delà d'une dizaine d’éléments on va plus vite avec un filtre/recherche/tri.

> indiquer plus clairement que toutes les références ne sont pas affichées ;

J'ai changé un peu le visuel sur une autre branche :

![image](https://user-images.githubusercontent.com/333276/122370461-5d5ab300-cf5f-11eb-891c-aba7b214b21b.png)

Qu'est-ce que tu en penses ?

> proposer une option pour dérouler les autres références plutôt que rechercher : je ne sais pas forcément quelles sont mes références, donc je ne sais pas quoi chercher.

Comment est-ce que tu retrouves une référence bibliographique dans Zotero ? Est-ce qu'il ne serait pas préférable d'afficher le "Title", "Author" et "Date" plutôt que l'identifiant ? On peut aussi faire une recherche sur plusieurs champs (actuellement on cherche/filtre uniquement sur l'identifiant).

Cliquer pour afficher plus de résultats me semble pénible, si la référence bibliographique est dans la troisième ou quatrième page c'est rapidement frustrant.

lakonis commented:
Je suis d'accord avec les suggestions de @antoinentl et je commente ci-dessous les réponses de @Mogztter 

> On peut augmenter un peu la limite mais je pense qu'au delà d'une dizaine d’éléments on va plus vite avec un filtre/recherche/tri.

Je suis d'accord avec toi, mais ca n'est que personnel. C'est-à-dire qu'on s'adresse à des personnes n'ayant pas forcément cette pratique de l'autocomplétion, ou du champ de recherche filtrant. Il faut donc les prendre en compte.

> Qu'est-ce que tu en penses ?

Ca me semble très bien.

> Comment est-ce que tu retrouves une référence bibliographique dans Zotero ? Est-ce qu'il ne serait pas préférable d'afficher le "Title", "Author" et "Date" plutôt que l'identifiant ? On peut aussi faire une recherche sur plusieurs champs (actuellement on cherche/filtre uniquement sur l'identifiant).

Tu as raison. Est ce que c'est jouable sur cette interface d'afficher ces 3 champs en 3 colonnes ? Si déjà la requête vient interroger title/author/date (tout en n'affichant que les identifiants), alors ça sera vraiment un plus. Proposition dans ce sens: lors d'une requête, afficher les résultats en listant les identifiants et à côté ou juste dessous un aperçu du contexte où la chaine de caractères a été trouvée.

Ex: 
- requête: `misé`
- résultats: `@hugo_les_1862 : Les **Misé**rables`

Mmellet commented:
Je suis d'accord avec le point de @antoinentl : il me semble que le déploiement de l'entièreté de la bibliographie dans le volet bibliographie doit demeurer une option : selon les pratiques, le filtre pourra très bien être utilisé mais nous avons aussi des auteurs qui "débarquent" sur stylo et intuitivement ils ne vont pas comprendre qu'ils doivent utiliser la barre de recherche pour atteindre par exemple les références non affichées par défaut... 

Pour ce qui est de l'affichage des références, si on décide d'afficher le titre, l'auteur et la date plutôt que l'identifiant, je propose alors d'inverser le systèmet actuellement en place : au passage du curseur sur la référence, on aurait l'identifiant bib. Je propose cela car selon les cas des références que l'on va avoir, avec le titre, l'auteur et la date affichés, on va avoir des doublons pour le cas par exemple de volumes cités (mais ça restera des cas rares, je pense).

thom4parisot commented:
J’ai l’impression que le premier petit pas serait de procéder au changement technique sans lui lier le changement d’interface.

ggrossetie commented:
@oncletom En fait c'est assez compliqué car tout est un peu lié. Je pense qu'il faudrait qu'on ait des fonctions plus unitaires. Par exemple, actuellement, quand on enregistre les références bibliographiques on enregistre une nouvelle version avec l'ensemble des informations de l'article en repassant par le composant `Write`, ce qui crée une dépendance.

Je pense que le composant "Bibliographie" devrait gérer l'enregistrement des données et ensuite publier un événement pour rafraîchir l'éditeur.

Au final, il y aura quand même pas mal de changements même si on conserve la même interface.

antoinentl commented:
Je commente ici, si nécessaire je déplacerai.

Voici les principes que nous retenons :

- 25 références affichées maximum dans la barre latérale, avec un bouton/lien "voir plus de références" pour en afficher plus ;
- un petit champ de recherche pour chercher et afficher des références, toujours dans le panneau latéral ;
- pour chaque référence dans le panneau latéral il faut afficher :
    - sur une première ligne : le premier auteur·trice, le titre (40 premiers caractères suivis de 3 petits points), la date
    - sur une seconde ligne, moins en évidence : la clé bibtex
- de l'autocomplétion dès que l'on tape `@` _dans_ le texte.

Est-ce que ces points sont clairs ?

thom4parisot commented:
C'est bon pour moi.

ggrossetie commented:
> 25 références affichées maximum dans la barre latérale, avec un bouton/lien "voir plus de références" pour en afficher plus ;

On est d'accord que lorsque l'on clique sur "voir plus de références" on affiche l'ensemble des références (et non les 25 références suivantes) ?
Cela me fait penser qu'il faudrait statuer sur les libellés dans l'interface. Globalement, l'interface est à 90% en anglais mais il y a quelques libellés en français (notamment dans le menu "Metadata").
Si on veut rester sur de l'anglais, alors le libellé sera "See more bibliographic references" ? ou "Show all bibliographic references" ? Est-ce qu'on peut raccourcir en "Show all references" ?

> un petit champ de recherche pour chercher et afficher des références, toujours dans le panneau latéral ;

On applique une recherche sur l'ensemble des champs affichés dans les résultats ? (i.e., le premier auteur·trice, le titre, la date, la clé bibtex).
Si oui, il faudra surement mettre en surbrillance le texte qui correspond à la recherche. 

> le titre (40 premiers caractères suivis de 3 petits points), 

Si la recherche trouve une correspondance sur le titre au delà des 40 premiers caractères, comment l'indiquer visuellement ? Est-ce qu'on peut mettre en surbrillance les "..." ?

> de l'autocomplétion dès que l'on tape @ dans le texte.

@antoinentl C'est un sujet assez conséquent, je pense qu'il est préférable de créer une issue GitHub dédiée.

ggrossetie committed:
Affichage détaillée des références

- Barre de recherche permettant de filter/rechercher des références
- Affiche 25 références bibliographique par défaut
- Ajout d'un bouton permettant d'afficher l'ensemble des références bibliographique
- Affiche le nom de l'auteur/autrice, le titre, l'année et la clé bibtex pour chaque référence bibliographique

ggrossetie commented:
#### Ce qu'on a fait 

- 25 références affichées maximum, avec un bouton "Show all {total} references" pour afficher l'ensemble des références
- un champ de recherche pour chercher des références (la recherche s'applique uniquement sur lé clé bibtex)
- pour chaque référence, on affiche :
  * sur une première ligne : le premier auteur·trice, le titre (40 premiers caractères suivis de 3 petits points), la date
  * sur une seconde ligne, moins en évidence : la clé bibtex

#### Liste les 25 références (pas de recherche/filtre actif)

![image](https://user-images.githubusercontent.com/333276/141085580-6231119a-1cf0-4bea-aefc-04cfd32bd48c.png)

#### Bouton permettant d'afficher l'ensemble des références bibliographiques

![image](https://user-images.githubusercontent.com/333276/141085652-9b1354d9-981f-40b9-a314-a6b3812a5dc9.png)

#### Recherche/filtre actif

![image](https://user-images.githubusercontent.com/333276/141085739-9a30ab08-d4f6-40ce-8b50-782c049cbfcc.png)

ggrossetie commented:
>  C'est un sujet assez conséquent, je pense qu'il est préférable de créer une issue GitHub dédiée.

@antoinentl c'est déjà le cas (https://github.com/EcrituresNumeriques/stylo/issues/271) et ça sera traité dans une autre pull request

ggrossetie commented:
> On applique une recherche sur l'ensemble des champs affichés dans les résultats ? (i.e., le premier auteur·trice, le titre, la date, la clé bibtex).
Si oui, il faudra surement mettre en surbrillance le texte qui correspond à la recherche.
Si la recherche trouve une correspondance sur le titre au delà des 40 premiers caractères, comment l'indiquer visuellement ? Est-ce qu'on peut mettre en surbrillance les "..." ?

@antoinentl Pour rappel, actuellement, on recherche/filtre uniquement sur la clé bibtex. Si il faut faire une évolution, je te laisse créer une issue avec le comportement souhaité.

Commit:
Extract code as `deriveAuthorNameAndDate()`, add it into `toEntries()` and test it

Commit:
It sounds the test was incorrect in the first place, and never ran again because the file was named `.mjs`

Commit:
Use `<span>` tags as descendants of `<p>` instead of `<div>` (not semantically correct)

Commit:
Maintain button border when in disabled state

ggrossetie commented on front/gatsby/src/components/Write/ReferenceList.jsx:
```diff
@@ -1,16 +1,38 @@
+  if (filter) {
+    bibTeXFound = articleBibTeXEntries
+      .filter((entry) => entry.key.toLowerCase().indexOf(filter.toLowerCase()) > -1)
+      //.slice(0, 10)
```
```suggestion
```

ggrossetie commented on front/gatsby/src/components/Write/ReferenceList.jsx:
```diff
@@ -1,16 +1,38 @@
 
-function ReferenceList ({ articleBibTeXEntries }) {
+function ReferenceList({ articleBibTeXEntries }) {
+  // state showAll boolean (default false)
```
```suggestion
```

ggrossetie commented on front/gatsby/src/helpers/bibtex.test.js:
```diff
@@ -5,14 +5,14 @@ describe('parse', () => {
             title = {test19
         }
-        
+        bib2key
```
🤔

ggrossetie committed:
Supprime commentaire en trop

antoinentl commented:
@Mogztter @oncletom C'est de toute beauté, bravo !
Je laisse @Mmellet et @marviro jeter un œil, pour moi c'est ok pour être déployé en production **le 17 novembre** sauf avis contraire _avant_ cette date.

Mmellet commented:
C'est ok pour être déployé de mon côté, 

j'ai quelques questions (notamment à discuter aec @antoinentl ) peut-être pour le futur : 

- est-ce qu'on donne l'option de revenir à l'affichage non-exhaustif ? 
- est-que pour être cohérent avec le bibtex et pour facilier la lecture des références si elles sont par ordre alphabétique : on ferait pas plutôt mieux de mettre dans la référence Nom Prénom | Titre ? 

Bravo en tout cas c'est très fluide !
